components/support/sync15-traits/src/store.rs

```diff
@@ -0,0 +1,93 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::{
+    client::ClientData, telemetry, CollectionRequest, Guid, IncomingChangeset, OutgoingChangeset,
+    ServerTimestamp,
+};
+use failure::Error;
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct CollSyncIds {
+    pub global: Guid,
+    pub coll: Guid,
+}
+
+/// Defines how a store is associated with Sync.
+#[derive(Debug, Clone, PartialEq)]
+pub enum StoreSyncAssociation {
+    /// This store is disconnected (although it may be connected in the future).
+    Disconnected,
+    /// Sync is connected, and has the following sync IDs.
+    Connected(CollSyncIds),
+}
+
+/// Low-level store functionality. Stores that need custom reconciliation logic
+/// should use this.
+///
+/// Different stores will produce errors of different types.  To accommodate
+/// this, we force them all to return failure::Error.
+pub trait Store {
+    fn collection_name(&self) -> std::borrow::Cow<'static, str>;
+
+    /// Prepares the store for syncing. The tabs store currently uses this to
+    /// store the current list of clients, which it uses to look up device names
+    /// and types.
+    ///
+    /// Note that this method is only called by `sync_multiple`, and only if a
+    /// command processor is registered. In particular, `prepare_for_sync` will
+    /// not be called if the store is synced using `sync::synchronize` or
+    /// `sync_multiple::sync_multiple`. It _will_ be called if the store is
+    /// synced via the Sync Manager.
+    // TODO: This is pretty cludgey and will be hard to extend for any case other
+    // than the tabs case. We should find another way to support tabs...
```

Comment:
I left some comments about this on #2562, but this feels like the best approach now. I'm really regretting adding the clients engine hack to support just tab syncing—which was itself a hack to accommodate syncing without a clients engine, which we don't really support anymore thanks to Megazording and the sync manager. Let's get an issue on file to clean this up and figure out how to remove `prepare_for_sync` eventually.

Reply:
Filed #2590, and referenced in the comment.